feat: markdown upload media

pauldambra opened the pull request:
## Problem

You can add gifs and images to text cards but need to know how to write markdown (and realise it is possible)

## Changes

* Adds an API (authenticated) to allow image upload
* Adds an endpoint to view images (immutable cache headers set)
* Adds some basic validation
* Adds UI to allow drop of file onto a text card (well, any component using the `LemonTextMarkdown`) to upload the image and insert a link to it in the markdown content

![dancing-banana](https://user-images.githubusercontent.com/984817/195578167-a9e0a544-8a77-4104-8b3e-c6e0e4b153a9.gif)


## Follow-up required

* Need to add analytics on use of text cards and images
* Need to write a docs page for text cards

## How did you test this code?

Adding developer tests and testing it locally

pauldambra committed:
wip

pauldambra committed:
remove some css

pauldambra committed:
add API to upload and retrieve media

pauldambra committed:
semi-working hacky flow

pauldambra committed:
wireup front end

pauldambra committed:
a little validation

pauldambra committed:
don't allow upload if object storage is not available

pauldambra committed:
add some API docs

mariusandra commented on frontend/src/lib/components/LemonTextArea/LemonTextArea.tsx:
```diff
@@ -63,13 +69,96 @@ interface LemonTextMarkdownProps {
 export function LemonTextMarkdown({ value, onChange, ...editAreaProps }: LemonTextMarkdownProps): JSX.Element {
+    const { objectStorageAvailable } = useValues(preflightLogic)
+
+    const [drag, setDrag] = React.useState(false)
+    const [isUploading, setIsUploading] = React.useState(false)
```
Why `useEffect`, but `React.useState`?

pauldambra replied:
* updates front end coding conventions

![side-eye-monkey](https://user-images.githubusercontent.com/984817/195589755-2bfe7806-8267-4935-9453-fe2374cc2fae.png)

Twixes replied:
Only the first 64 bits of a UUIDT are incremented-ish (the first 48 are the timestamp for smoother indexing in DBs), the other 64 bits are securely random – so this should be unguessable.

mariusandra replied:
So if you get the timestamp right, you have 1,208,925,819,614,629,273,600,000 guesses to get all possible files. Probably fine then.

mariusandra reviewed (commented):
Looks sweet! Some small questions:

1. I went all metahog and added a very large gif on a dashboard.

![2022-10-13 13 29 14](https://user-images.githubusercontent.com/53387/195585119-09f9dfc3-7666-492e-9ae1-e4c4c917f0ad.gif)

I think markdown images should have "max-width: 100%" set, as who knows what kind of high-dpi content gets uploaded by kids these days.

2. Are UUID-s secure (aka random) enough to act as links? Could I reasonably guess someone else's dashboard files given enough (unmonitored) tries? All these UUIDT-s do look incremental to some degree. So should we use some random 64-character alphanumeric string instead?

3. Is this instrumented well enough that you can create a dashboard tracking how much this feature is used? 😁

pauldambra committed:
pick one way and use it

pauldambra committed:
fix a bunch of things

pauldambra committed:
add file name into url

pauldambra committed:
gzip uploaded media responses

pauldambra committed:
remove console log

pauldambra commented:
okey dokey @mariusandra 

* added gzip to end point
* updated other bits from review

The UUIDs are being used as DB ids so if they aren't unique enough in a model namespace then we have other problems 🤣 But there's no harm (probably) in using the filename as well as the ID. I'll double check there's no validation needed on the content of the filename before we release this past our team

pauldambra commented:
Have statsd counters for the API and events when upload succeeds or fails

<img width="1113" alt="Screenshot 2022-10-13 at 13 13 04" src="https://user-images.githubusercontent.com/984817/195595118-6e8a4e01-36ee-45d9-bc29-091c9f290178.png">

mariusandra reviewed (approved):
Very well then 👍 

If I get a `max-width: 100%` for the images, I'm happy.

![2022-10-13 14 32 54](https://user-images.githubusercontent.com/53387/195597323-22cc9af3-889a-4e54-a72c-e1390b209415.gif)

Twixes reviewed (dismissed):
This is actually really cool! Found a few rough edges though

pauldambra commented on posthog/api/uploaded_media.py:
```diff
@@ -0,0 +1,95 @@
+            else:
+                raise UnsupportedMediaType(file.content_type)
+        except KeyError:
+            return Response("file missing.", status=status.HTTP_400_BAD_REQUEST)
```
I parsed it as, the resource exists at /api/projects/\d+/uploaded_media but the data passed to it is incorrect and shouldn't be repeated as it is.

pauldambra replied:
Switched it for `raise BadRequest` with a slightly less lazy message. So it's clearer for the future traveller

pauldambra commented on frontend/src/lib/components/LemonTextArea/LemonTextArea.tsx:
```diff
@@ -63,13 +69,96 @@ interface LemonTextMarkdownProps {
+    const [isUploading, setIsUploading] = React.useState(false)
+
+    const dropRef = React.createRef<HTMLDivElement>()
+    let dragCounter = 0
```
🤯 

```
    // dragCounter and drag are used to track whether the user is dragging a file over the textarea
    // without drag counter the textarea highlight would flicker when the user drags a file over it
```

Github Copilot wrote almost all of the comment, how?!

pauldambra committed:
comment on why dragcounter

pauldambra committed:
guide user to the feature

pauldambra committed:
extract constant

pauldambra committed:
more clearly a bad request

pauldambra commented on frontend/src/lib/components/LemonTextArea/LemonTextArea.tsx:
```diff
@@ -63,13 +69,96 @@ interface LemonTextMarkdownProps {
 }
 
 export function LemonTextMarkdown({ value, onChange, ...editAreaProps }: LemonTextMarkdownProps): JSX.Element {
+    const { objectStorageAvailable } = useValues(preflightLogic)
```
## object storage disabled

<img width="495" alt="Screenshot 2022-10-13 at 16 53 49" src="https://user-images.githubusercontent.com/984817/195646487-2efbf0aa-09b5-4777-9cb3-954932956eb3.png">

## object storage enabled

<img width="499" alt="Screenshot 2022-10-13 at 16 51 23" src="https://user-images.githubusercontent.com/984817/195646566-c0e6ff39-7407-4802-95b9-65c6258bc2e6.png">

Twixes replied:
Cool! Could we just make it so that the text fits into one line? Two draw a bit too much attention to some extra info. For the drag-and-drop-on case it could just say "Attach images by dragging and dropping them", if unavailable… The message might not be necessary? The question is, in what situations is object storage unavailable (besides not having the `object_storage` service running in local dev)? If this is only when something is misconfigured, we can be pretty verbose; if it's common and totally acceptable, a big "Cannot upload ..." message is going to be annoying.

BTW "Markdown" should be capitalized, it's a proper noun unlike "markup" 😅)

pauldambra replied:
got them both to one line and moved object storage note into a tooltip

<img width="498" alt="Screenshot 2022-10-13 at 18 09 17" src="https://user-images.githubusercontent.com/984817/195661983-9e7a8f86-ab41-4128-b290-703fc187340d.png">
<img width="670" alt="Screenshot 2022-10-13 at 18 07 54" src="https://user-images.githubusercontent.com/984817/195661987-ee7e591d-0602-456c-8ccf-6ec9d3d8a9bc.png">

Twixes commented on frontend/src/lib/components/LemonTextArea/LemonTextArea.tsx:
```diff
@@ -63,13 +68,118 @@ interface LemonTextMarkdownProps {
+                >
+                    <LemonTextArea {...editAreaProps} autoFocus value={value} onChange={onChange} />
+                    <div className="text-muted inline-flex items-center space-x-1">
+                        <IconMarkdown className={'text-2xl'} />
+                        <span>Markdown formatting support</span>
+                    </div>
+                    {objectStorageAvailable ? (
+                        <div className="text-muted inline-flex items-center space-x-1">
+                            <IconUploadFile className={'text-2xl mr-1'} />
```
I suppose `IconMarkdown` should have `mr-1` too

pauldambra replied:
The width of the Markdown icon made the text look off to me. So `mr` on the two other icons to get the text closer to lining up

<img width="498" alt="Screenshot 2022-10-13 at 18 09 24" src="https://user-
images.githubusercontent.com/984817/195662081-a65fe30e-e540-40e8-8ea6-da03c53a95e6.png">

pauldambra committed:
text fangling

pauldambra committed:
not bad request

pauldambra committed:
not bad request
